### DISCUSSION TITLE
How will code modularity be achieved in Cito?

### DISCUSSION THREAD

**Discussion by danon** [pfusik/cito#65] [Q&A]
You mentiond that Ćito isn't a general purpose language, but is more of a mean to express ideas and libraries, which can then be transpiled and used in other applications, I think that idea is pretty cool.

But I wonder, obviously if one is to write a library in cito, one must split it into multiple files for readability and reusability and maintainability. Will this be achieved like Java classloader/Php autoload? Or more of a python `import`? Or some other way?

And when transpiled, will it result in multiple files in the output language (like multiple `.java` files) or one humongous monolith?

**Comment by pfusik** [comment]
You can split Ć code into source files much like you split C#. That is, one class per file, or several classes per file, or even your whole library in one source file.

This in no way affects the effect of the transpilation. Java requires every `public class` to be in separate source file, so this is how Java code gets generated. The Java source files could be then compiled and packed as a `jar`, but there is currently no dedicated tooling for that.

I don't know PHP, specifically what's "autoload".

**Reply by danon** [reply]
That's not what I meant.

My question was how exactly the `.ci` files reference one another?

In Java (with class loader), also in Kotlin and PHP's autoloader, it works like first all classes are loaded into memory at first, and then they are always available in your scope. You can use `import` in Java and `use` in PHP, to shorten the namespace, if you'd like, but that's fine to do just `new org.something.something.MyClass();` in Java or `new \something\something\MyClass();` in PHP. Same is for C#.

However in Python, JavaScript (and I also think in C++) files aren't loaded all at once at the start, but you have to do something to load it. For example in Python you have to do `from random import randrange` and in JavaScript `import React from "react";`. 

In other words, if you add class in Java/c#/Php (if you have autoloader) you can't do anything not to load a class, they're always loaded. 

---

However looking at the fact that in `.ci` one file per class, then I assume it's the first option?

BTW.

How do I write it in Ć? Do I use `import`/`use` or something else?

**Reply by pfusik** [reply]
`.ci` files are listed on the `cito` command line and transpiled as a whole. You reference classes (whether they are in the same file or another) just by the class name.

Example:
`Foo.ci`:
```csharp
static class Foo
{
    public static void Bar()
    {
        Console.WriteLine("Hello, world!");
    }
}
```

`Hello.ci`:
```csharp
public static class Hello
{
    public static void Greet()
    {
        Foo.Bar();
    }
}
```

Command line:

    cito -o Hello.java Foo.ci Hello.ci

produces `Hello.java` *and* `Foo.java`.

To specify target Java package, use the `-n` cito option.

**Reply by danon** [reply]
So it's similar to Java/C#/Php autoload. Simply all the classes are available everywhere?

Are there namespaces?

**Reply by pfusik** [reply]
There are no namespaces in Ć source. There's a discussion on it: #6. I'm open for ideas on that!
